Smoke testing for Qemu

thenor opened the pull request:
Adds a script for smoke testing triggered from the Jenkinsfile.

Signed-off-by: Therese Nordqvist <therese.nordqvist@pelagicore.com>

erikboto commented on Jenkinsfile:
```diff
@@ -68,6 +68,13 @@ void buildManifest(String variant_name, String bitbake_image) {
             }
         }
 
+        if ("smoke-test" in options || "export-oe-tests" in options) {
```
In another project we use a Set called "options" which is passed as a parameter to buildManifest. That Set it not available here and that's what the build is failing on.

thenor replied:
Thanks! I realized that this is why is fails, but was trying to find "options". Do you have some time for help me, @erikboto? @rpannek started to help me but is not in the office now.

erikboto commented on Jenkinsfile:
```diff
@@ -91,6 +98,75 @@ void buildManifest(String variant_name, String bitbake_image) {
             }
         }
 
+        if ("smoke-test" in options) {
```
Same issue as line 71.

tobsan commented on Jenkinsfile:
```diff
@@ -91,6 +98,77 @@ void buildManifest(String variant_name, String bitbake_image) {
         }
 
+        if (variant_name == "qemu-x86-64_nogfx") {
+            try {
```
Perhaps all this could be a separate Jenkinsfile that we load and run if needed? Since it is such a different mode of operation I think that would make sense.

thenor replied:
That sounds like a good idea. For now this is only a WIP, but I will break it out to finish the user story.

thenor commented:
@erikboto Maybe I can give you access to my repo so you can push your changes to this branch and PR?

erikboto commented on test-scripts/local.conf.appendix:
```diff
@@ -0,0 +1,12 @@
+    ping \
+    ssh \
+    auto \
+"
```
This file can be removed right, there's another version with the same name that I think is the one used.

thenor replied:
Oops. A merge mistake.

thenor replied:
My idea here was to have one file for the test specific env. Maybe it's overkill? What do you think @tobsan (or others)?

rpannek replied:
A more specific one is nice but I guess only one should be used?

thenor replied:
There is configuration needed for the tests, but not for anything else. That was shy it felt most reasonable for me to have a new conf file for the tests.

erikboto commented on local.conf.appendix:
```diff
@@ -1,3 +1,20 @@
 
+# Needed since a bbappend in meta-ivi adds libusb PACKAGECONFIG which is 
+# broken for some weird reason
+BBMASK .= "|meta-ivi/recipes-devtools/qemu/"
```
Maybe this should be done not only when testing? So moved to the meta-pelux/variant/common/local.conf?

thenor replied:
@tobsan what do you think?

tobsan replied:
If it is a generic setting that we always want to have enabled, then it should of course go in ``meta-pelux/variant/common/local.conf``

GordanM replied:
Furthermore, shouldn't there be a change upstream to meta-ivi that fixes this `PACKAGECONFIG`? I guess it would be nice to inform them about this. Does the downstream project experience these side effects as well?

thenor replied:
@GordanM You were removing a bunch of qemu stuff. This is still ok to include in our general config?

thenor replied:
I guess there are two topics being discussed now. The upstreaming and solving of the qemu devtool from meta-ivi and if we should include this change in the local.conf now.

GordanM replied:
@thenor I've removed some meta-bistro QEMU startup scripts that are not being used anywhere, should not impact any functionality. 

Nice! That is good to know. We can always create two separate changes for this, maybe this even belongs in the temp patches then? You know best, I've just wanted to point this out :).

thenor commented on local.conf.appendix:
```diff
@@ -1,3 +1,20 @@
+# Needed since a bbappend in meta-ivi adds libusb PACKAGECONFIG which is 
+# broken for some weird reason
+BBMASK .= "|meta-ivi/recipes-devtools/qemu/"
+
```
This and the rest of the added lines I will keep only in the test-scripts directory.

thenor commented:
Is there a way to make this build as if it was the nightly? @tobias, do you know? Or should I wait for the next nightly after it is merged to see that it passes?

thenor commented:
I got the connection refused issue on the rpi build, but all other images look fine.

GordanM commented:
I believe this is good enough right? Because there is no impact on the rpi image :)

rpannek commented on ci-scripts/yocto.groovy:
```diff
@@ -87,6 +95,25 @@ void buildImageAndSDK(String yoctoDir, String imageName, boolean dev=true) {
+        stage("Perform smoke testing") {
+            vagrant("/vagrant/cookbook/yocto/runqemu-smoke-test.sh ${yoctoDir} ${imageName}")
+        }
+    } catch(e) {
```
I really dislike this way of catching everything and then not even printing what actually happened because this makes impossible to find what actually happened. Either remove the catch-all or make it more specific for some type of a error and print the call stack or dive some possibility to find out what happened instead of just printing a string.

thenor replied:
See the newly added code below, @rpannek. Is this what you had in mind?

rpannek commented on local.conf.appendix:
```diff
@@ -1,3 +1,6 @@
+# Needed since a bbappend in meta-ivi adds libusb PACKAGECONFIG which is
+# broken for some weird reason
+BBMASK .= "|meta-ivi/recipes-devtools/qemu/"
```
Did we report a bug?

thenor replied:
Not yet. I will discuss this with @frznlogic and see if we can find a way forward (either we find the solution or we just report it as a bug).

thenor replied:
I have talked to @frznlogic and we will look into this. It will not be done within this PR though.

rpannek replied:
Sure, just so we report the bug at least.

frznlogic replied:
sorry, I missed this discussion earlier. I did look at it briefly, the PACKAGECONFIG is super old... I'll remove it and see what people/the build system think of it.

thenor commented:
@rpannek @GordanM What do you think about keeping the Qemu flag that I introduced in the last commit. Should I keep it or remove it?

rpannek commented on Jenkinsfile:
```diff
@@ -10,34 +10,56 @@
-            checkout scm
-            def manifests = load "ci-scripts/yocto.groovy"
-            manifests.buildManifest("intel", "core-image-pelux-minimal")
+            boolean only_qemu = env.QEMU == "true"
```
That is a really weird way of implementing that by copy&pasting it that many times.

I would not try to do a special thing for qemu, but instead abstract it away so it is easy to extend for example you could create this map programatically before the parallel() call. And there you could have a list of the jobs which are default (which would be all of them) and then prepare so one can manipulate this list to remove some of them, or overwrite it with only qemu in it on demand.

This way you would make it possible to just run the rpi build ord only the qemu and intel-qtauto, or any other combination without every time adding special if clauses.

thenor replied:
I guess you mean this as an improvement to the Jenkinsfile, so not actually within the scope of adding smoke tests. It does make sense to me to try to get that to work.

thenor replied:
I can remove the qemu flag. I just wanted to use it for being able to run from Jenkins in a simple way without hard-coding.

thenor replied:
Sorry. I misunderstood your comment. Yes, I agree this is not a nice way of doing this.

thenor commented:
@rpannek I removed the flag I had introduced and we will introduce a nicer way to do this using maybe a list sent from Jenkins, but not included in the scope of this pull request.
